api/desecapi/models.py

```diff
@@ -313,3 +369,205 @@ class RR(models.Model):
 
     def __str__(self):
         return '<RR %s>' % self.content
+
+
+class AuthenticatedAction(models.Model):
+    """
+    Represents a procedure call on a defined set of arguments.
+
+    Subclasses can define additional arguments by adding Django model fields and must define the action to be taken by
+    implementing the `act` method.
+
+    AuthenticatedAction provides the `mac` property that returns a Message Authentication Code (MAC) based on the
+    state. By default, the state contains the action's name (defined by the `name` property) and a timestamp; the
+    state can be extended by (carefully) overriding the `mac_state` method. Any AuthenticatedAction instance of
+    the same subclass and state will deterministically have the same MAC, effectively allowing authenticated
+    procedure calls by third parties according to the following protocol:
+
+    (1) Instantiate the AuthenticatedAction subclass representing the action to be taken with the desired state,
+    (2) provide information on how to instantiate the instance and the MAC to a third party,
+    (3) when provided with data that allows instantiation and a valid MAC, take the defined action, possibly with
+        additional parameters chosen by the third party that do not belong to the verified state.
+    """
+    created = models.PositiveIntegerField(default=lambda: int(datetime.timestamp(datetime.now())))
+
+    class Meta:
+        managed = False
+
+    def __init__(self, *args, **kwargs):
+        # silently ignore any value supplied for the mac value, that makes it easier to use with DRF serializers
+        kwargs.pop('mac', None)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def name(self):
```

Comment:
Why is this a property and not an attribute?

Also, it would shadow any payload of the same name, right? (Should it be .Meta.name?)

Reply:
It's a property so that it can be an "abstract attribute". Also, yes, shadows payload of the same name, but that's not a problem now.